builder/azure/arm/template_factory.go

```diff
@@ -30,6 +52,20 @@ func GetKeyVaultDeployment(config *Config) (*resources.Deployment, error) {
 	return createDeploymentParameters(*doc, params)
 }
 
+func GetCustomScriptDeployment(config *Config) (*resources.Deployment, error) {
+	params := &template.TemplateParameters{
+		CommandToExecute: &template.TemplateParameter{Value: config.CustomScript},
+		VMName:           &template.TemplateParameter{Value: config.tmpComputeName},
+	}
+
+	builder, err := template.NewTemplateBuilder(template.CustomScript)
+	if err != nil {
+		return nil, err
+	}
+	doc, _ := builder.ToJSON()
+	return createDeploymentParameters(*doc, params)
+}
+
 func GetVirtualMachineDeployment(config *Config) (*resources.Deployment, error) {
```

Comment:
have you considered having two templates - one for vm without custom script and one for with custom script and just doing one template deployment. Here the VM will be deployed and you will do another template deployment to install extension and that might get deplayed (and thus kind of defeating the purpose). Baking it into the same deployment template will let azure try and do the vmextension deployment as soon as it can post vm startup.

Reply:
Good call - that made the code so much simpler, and faster too. Success 💪